### PR TITLE
Add Proof size limit to sync client

### PR DESCRIPTION
**Pull request by dboehm-avalabs**
Since values can be arbitrarily long, proofs could end up larger than the allowed network message size.  Add a mechanism to  limit proofs to a reasonable size.

### CODE REVIEW

File: x/sync/network_server.go [hunk 1]
```diff
@@ -183,30 +257,91 @@ func (s *NetworkServer) HandleRangeProofRequest(
...
-			return nil // dropping request
+			return nil
+		}
+
+		bytesEstimate += getBytesEstimateOfProofNodes(rangeProof.EndProof)
+
+		// shrink more if the early keys are extremely large
+		for keyIndex := uint16(1); keyIndex < keyLimit; keyIndex++ {
+			nextKV := rangeProof.KeyValues[keyIndex]
+			kvEstBytes := len(nextKV.Key) + len(nextKV.Value) + binary.MaxVarintLen64
```
Comment (darioush) [inline x/sync/network_server.go:318]: should this be `2*binary.MaxVarintLen64`?

File: x/sync/network_server_test.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+				return
+			}
+			require.NotNil(proofResult)
+			if test.expectedResponseLen > 0 {
+				require.LessOrEqual(len(proofResult.KeyValues), test.expectedResponseLen)
+			}
+
+			bytes, err := merkledb.Codec.EncodeRangeProof(Version, proofResult)
+			require.NoError(err)
+			require.Less(len(bytes), int(test.request.BytesLimit))
```
Comment (darioush) [inline x/sync/network_server_test.go:130]: should this be LessOrEqual?

File: x/sync/network_server_test.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+			sender := common.NewMockSender(ctrl)
+			var proofResult *merkledb.RangeProof
+			sender.EXPECT().SendAppResponse(
+				gomock.Any(), // ctx
+				gomock.Any(),
+				gomock.Any(), // requestID
+				gomock.Any(), // responseBytes
+			).DoAndReturn(
+				func(_ context.Context, _ ids.NodeID, requestID uint32, responseBytes []byte) error {
+					// deserialize the response so we can modify it if needed.
```
Comment (darioush) [inline x/sync/network_server_test.go:102]: should we update this comment? eg, deserialize the response so we can store it in proofResult

File: x/sync/network_server_test.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+		it := trieDB.NewIteratorWithStart(deleteKeyStart)
+		if it.Next() {
+			err = view.Remove(context.Background(), it.Key())
+			require.NoError(err)
+		}
+		require.NoError(it.Error())
+		it.Release()
+
+		view.CommitToDB(context.Background())
```
Comment (darioush) [inline x/sync/network_server_test.go:173]: should we check the err here?

File: x/sync/network_server_test.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+	r := rand.New(rand.NewSource(1)) // #nosec G404
+
+	smallTrieDB, _, err := generateTrieWithMinKeyLen(t, r, defaultRequestKeyLimit, 1)
+	require.NoError(t, err)
+	smallTrieRoot, err := smallTrieDB.GetMerkleRoot(context.Background())
+	require.NoError(t, err)
+
+	tests := map[string]struct {
+		request             *RangeProofRequest
+		modifyResponse      func(*merkledb.RangeProof)
```
Comment (darioush) [inline x/sync/network_server_test.go:31]: should we remove `modifyResponse` since it's not used here?

File: x/sync/network_server_test.go [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+			sender := common.NewMockSender(ctrl)
+			var proofResult *merkledb.ChangeProof
+			sender.EXPECT().SendAppResponse(
+				gomock.Any(), // ctx
+				gomock.Any(),
```
Comment (darioush) [inline x/sync/network_server_test.go:252]: could we add a comment for this one too `// nodeID`?

File: x/sync/network_server.go [hunk 7]
```diff
@@ -136,32 +141,100 @@ func (s *NetworkServer) HandleChangeProofRequest(
...
-	limit := req.Limit
-	if limit > maxKeyValuesLimit {
-		limit = maxKeyValuesLimit
+	keyLimit := req.KeyLimit
+	if keyLimit > maxKeyValuesLimit {
+		keyLimit = maxKeyValuesLimit
 	}
+	bytesLimit := int(req.BytesLimit)
+	if bytesLimit > constants.DefaultMaxMessageSize {
+		bytesLimit = constants.DefaultMaxMessageSize
```
Comment (danlaine) [inline x/sync/network_server.go:150]: This should be lower than the max message size because if the number of bytes is equal to (or just below) the max message size then this message will be too large

File: x/sync/network_server.go [hunk 8]
```diff
@@ -183,30 +256,91 @@ func (s *NetworkServer) HandleRangeProofRequest(
...
+		zap.Stringer("req", req),
+		zap.Error(ErrMinProofSizeIsTooLarge),
+	)
+	return nil
+}
 
-	proofBytes, err := merkledb.Codec.EncodeRangeProof(Version, rangeProof)
-	if err != nil {
-		return err
+func getBytesEstimateOfProofNodes(proofNodes []merkledb.ProofNode) int {
```
Comment (danlaine) [inline x/sync/network_server.go:339]: I think this should probably live inside the codec -- we should try to keep the sync stuff as agnostic to the codec as possible

File: x/sync/network_server.go [hunk 9]
```diff
@@ -136,32 +141,100 @@ func (s *NetworkServer) HandleChangeProofRequest(
...
 
-	proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
-	if err != nil {
-		return err
+		bytesEstimate += getBytesEstimateOfProofNodes(changeProof.EndProof)
+		deleteKeyIndex := 0
+		changeKeyIndex := 0
+
+		// shrink more if the early keys are extremely large
+		for keyIndex := uint16(1); keyIndex < keyLimit; keyIndex++ {
```
Comment (danlaine) [inline x/sync/network_server.go:204]: Instead of adding key values / deleted keys until the proof is too large, would it make more sense to _delete_ key values until the proof is small enough?

File: x/sync/network_server.go [hunk 10]
```diff
@@ -136,32 +141,100 @@ func (s *NetworkServer) HandleChangeProofRequest(
...
-			return nil // dropping request
+			return nil
 		}
-		return err
-	}
 
-	proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
-	if err != nil {
-		return err
+		bytesEstimate += getBytesEstimateOfProofNodes(changeProof.EndProof)
```
Comment (danlaine) [inline x/sync/network_server.go:199]: Should we add some buffer to this to account for the fact that when we adjust the key limit, the new end proof may be larger than the current end proof?

File: x/sync/network_server.go [hunk 11]
```diff
@@ -183,30 +256,91 @@ func (s *NetworkServer) HandleRangeProofRequest(
...
+				zap.Error(ErrMinProofSizeIsTooLarge),
 			)
-			return nil // dropping request
+			return nil
+		}
+
+		bytesEstimate += getBytesEstimateOfProofNodes(rangeProof.EndProof)
+
+		// shrink more if the early keys are extremely large
+		for keyIndex := uint16(1); keyIndex < keyLimit; keyIndex++ {
```
Comment (danlaine) [inline x/sync/network_server.go:315]: Same comments that I left for change proof above apply here

File: x/sync/network_server.go [hunk 12]
```diff
@@ -136,32 +141,100 @@ func (s *NetworkServer) HandleChangeProofRequest(
...
+			// determine if the deleted key or changed key is the next smallest key
+			currentBytes := 0
+
+			// if there is a deleted key at deleteKeyIndex and
+			// (there are no more change keys or the changed key is larger than the deleted key)
+			if deleteKeyIndex < len(changeProof.DeletedKeys) &&
+				(changeKeyIndex >= len(changeProof.KeyValues) ||
+					bytes.Compare(changeProof.KeyValues[changeKeyIndex].Key, changeProof.DeletedKeys[deleteKeyIndex]) > 0) {
+				currentBytes = len(changeProof.DeletedKeys[deleteKeyIndex]) + binary.MaxVarintLen64
```
Comment (danlaine) [inline x/sync/network_server.go:213]: Instead of using `binary.MaxVarintLen64` can we export something from the codec so we're not assuming stuff about its implementation details here?

File: x/sync/network_server_test.go [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+				StartingRoot: startRoot,
+				EndingRoot:   endRoot,
+				KeyLimit:     defaultRequestKeyLimit,
+				BytesLimit:   2 * constants.DefaultMaxMessageSize,
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
```
Comment (danlaine) [inline x/sync/network_server_test.go:244]: We should make a new `require := require.New(t)` in each subtest so that if it errors, it'll include the name of the test that failed.

**Review by danlaine — CHANGES_REQUESTED** [review]

File: x/merkledb/history.go [hunk 14]
```diff
@@ -97,27 +97,32 @@ func (th *trieHistory) getValueChanges(startRoot, endRoot ids.ID, start, end []b
...
-				lastStartRootChange = item
-				return false
-			}
-			return true
-		},
-	)
-
-	// There's no change resulting in [startRoot] before the latest change resulting in [endRoot].
-	if lastStartRootChange == nil {
+	lastStartRootChange, ok := th.lastChanges[startRoot]
```
Comment (danlaine) [inline x/merkledb/history.go:100]: Is this change related to the ones in the sync package? If not can we do this in another PR?

Comment (dboehm-avalabs) [reply x/merkledb/history.go:100]: https://github.com/ava-labs/avalanchego/pull/1310

File: x/sync/network_server.go (hunk 10 above)
Comment (dboehm-avalabs) [reply x/sync/network_server.go:198]: What amount of buffer do you think would be sufficient?  Theoretically the new end proof could be arbitrarily larger than the current end proof.

File: x/sync/network_server.go (hunk 9 above)
Comment (dboehm-avalabs) [reply x/sync/network_server.go:203]: it would require me to already know the total size, so I would have to add up all the values first.

File: x/sync/network_server.go (hunk 7 above)
Comment (dboehm-avalabs) [reply x/sync/network_server.go:149]: I'm fine moving this to be a different constant.  What do you think an appropriate offset would be?

Comment (danlaine) [reply x/sync/network_server.go:149]: The max size of the proof should be `[max avalanchego message size] - [app response overhead size]` but it's not entirely clear to me how to compute the latter given that the message is proto. To be safe maybe we could say  32 KiB? I doubt it's actually anywhere near that large but better safe than sorry, and I suspect that it won't matter in most cases since the proof won't be near the max size. We should probably put a TODO to return to this question

File: x/sync/network_server.go (hunk 9 above)
Comment (danlaine) [reply x/sync/network_server.go:201]: >We are eliminating half the key/values before we get here

The current heuristic is:

```
for {
    // Get proof
    // If proof small enough, return
    // Halve max number of keys
    // Check if proof will still be too big even with number of keys halved. If so, reduce number of keys more.
}
```

What I'm wondering is: what if instead of halving the number of keys, we just iterate over the proof we got, removing the greatest keys until the proof is small enough?

File: x/sync/network_server.go [hunk 15]
```diff
@@ -136,32 +145,89 @@ func (s *NetworkServer) HandleChangeProofRequest(
...
+		if err != nil {
+			return err
+		}
+		if len(proofBytes) < bytesLimit {
+			return s.appSender.SendAppResponse(ctx, nodeID, requestID, proofBytes)
+		}
+		// the proof size was too large, try to shrink it
+
+		// ensure that the new limit is always smaller
+		keyLimit = uint16((len(changeProof.KeyValues) + len(changeProof.DeletedKeys)) / 2)
```
Comment (danlaine) [inline x/sync/network_server.go:185]: What's the rationale for doing this instead of just `keyLimit /= 2`?

Comment (dboehm-avalabs) [reply x/sync/network_server.go:185]: keyLimit/2 might be more keys than were actually returned, which means it will generate another proof that is too large.

File: x/sync/network_server.go (hunk 9 above)
Comment (dboehm-avalabs) [reply x/sync/network_server.go:201]: Having said all that, we could just do the key shaving and try to add some level of buffer and just hope for the best.

**Review by danlaine — APPROVED** [review]

**Review by darioush — APPROVED** [review]

File: x/sync/README.md [hunk 16]
```diff
@@ -0,0 +1,4 @@
+## TODOs
+
+- [ ] Handle errors on proof requests.  Currently, any errors that occur server side are not sent back to the client.
```
[the remark this answers is not in the document]
Comment (dboehm-avalabs) [reply x/sync/README.md:3]: I agree, but currently no errors of any sort are returned even if the request was valid.
